lib/spack/spack/paths.py

```diff
@@ -9,135 +9,318 @@
+        # locations)
+
+        old_gpg_path = os.path.join(self.prefix, "opt", "spack", "gpg")
+        if dir_is_occupied(old_gpg_path):
+            self.gpg_path = old_gpg_path
+        else:
+            self.gpg_path = os.path.join(self.user_cache_path, "gpg")
+
+        old_gpg_keys_path = os.path.join(self.var_path, "gpg")
+        if dir_is_occupied(old_gpg_keys_path):
+            self.gpg_keys_path = old_gpg_keys_path
+        else:
+            self.gpg_keys_path = os.path.join(self.user_cache_path, "gpg-keys")
+
+        self.modules_base = None
+        for module_dir in ["lmod", "modules"]:
+            if dir_is_occupied(os.path.join(self.share_path, module_dir)):
+                self.modules_base = self.share_path
+        if not self.modules_base:
+            self.modules_base = os.path.join(self.user_cache_path, "modules")
+
+        # ------ Next section
+        # Spack can also write data into the following locations, and their
+        # defaults are not controlled by SPACK/XDG_DATA_HOME or
+        # SPACK_USER_CACHE_PATH. Like the prior section, the data written
+        # into these locations isn't expected to take up much space, so in
+        # some cases defaults to "~" (in those cases in compliance with
+        # XDG defaults).
+
+        # There are three environment variables you can use to isolate spack from
+        # the host environment:
+        # - `SPACK_USER_CONFIG_PATH`: override `~/.spack` location (for config and caches)
+        # - `SPACK_SYSTEM_CONFIG_PATH`: override `/etc/spack` configuration scope.
+        # - `SPACK_DISABLE_LOCAL_CONFIG`: disable both of these locations.
+
+        #: User configuration location
+        self.user_config_path = os.path.expanduser(
+            os.getenv("SPACK_USER_CONFIG_PATH") or self.xdg_config_home
+        )
+
+        #: System configuration location
+        self.system_config_path = os.path.expanduser(
+            os.getenv("SPACK_SYSTEM_CONFIG_PATH") or os.sep + os.path.join("etc", "spack")
+        )
+
+        #: When Spack is provided by an admin to a user, the admin can
+        #: provide a config that only applies for the end-users
+        self.end_user_cfg_path = os.path.join(self.system_config_path, "end-user")
+
+        #: transient caches for Spack data (virtual cache, patch sha256 lookup, etc.)
+        #: overridden by `config:misc_cache`
+        self.default_misc_cache_path = os.path.join(
+            self.xdg_state_home, self.spack_instance_id, "misc-cache"
+        )
+
+        #: concretization cache for Spack concretizations
+        #: overridden by `config:concretization_cache:url`
+        self.default_conc_cache_path = os.path.join(self.default_misc_cache_path, "concretization")
+
+    def data_home_for_small_data(self):
+        if Location_vars.spack_data_home.value in os.environ:
+            return os.environ[Location_vars.spack_data_home.value]
+        else:
```

Comment:
```suggestion
            return os.environ[Location_vars.spack_data_home.value]
         elif self.xdg_state_home in os.environ:
            return os.path.expanduser(
                 os.path.join(os.environ[self.xdg_state_home], "spack", subdir)
             )
        else:
```

This might be useful if we expand the `SPACK_*` and `XDG_*` mapping. I think you can still do the small/large data mapping, but it would make it more consistent for the case where users are really into relying on `XDG_`.

The large/small data stands out to me as a case where the `SPACK_*` and `XDG_*` mapping bifurcation gets confusing. Adding this extra layer at least makes them more consistent.

Reply:
There are now `spack_` overrides for all XDG_ vars used in spack, so I think this bifurcation may be lessened. The large data home function is to highlight the need to distinguish between the case where the variable is explicitly defined (generally only important for _DATA_HOME).

This example appears to mix XDG_STATE_HOME and SPACK_DATA_HOME: was that intended?